File: c_plus_plus/strategy_pattern/MiniDuckSimulator.cpp
// Created by Randy Layne on 5/10/20.
//
#include "Squeak.h"
//

#include <iostream>
#include "Duck.h"
#include "FlyRocketPowered.h"
#include "MallardDuck.h"
#include "RubberDuck.h"
#include "ModelDuck.h"

using namespace std;
int main() {
    FlyBehavior *rocket = new FlyRocketPowered();

    Duck *ducks[3];

    ducks[0] = new MallardDuck();
    ducks[1] = new ModelDuck();
    ducks[2] = new RubberDuck();


    for (Duck *duck : ducks)
    {
        cout << *duck << endl;
        cout << duck->fly() << endl;
        cout << duck->quack() << endl << endl;
    }


    cout << "Poor rubber ducky can't fly. Let's change this behaviour by strapping on a rocket!\n";
    Duck *duck = ducks[2];

    FlyBehavior *f = new FlyRocketPowered();
    duck->setFlyBehavior(f);

    cout  << *duck << endl;
    cout << duck->fly() << endl;


    return 0;
}

File: strategy_pattern/java/com/randylayne/designpatterns/strategy/MiniDuckSimulator.java
package com.randylayne.designpatterns.strategy;

import java.util.ArrayList;
import java.util.List;

public class MiniDuckSimulator {
   public static void main(String[] args) {
      List<Duck> ducks = new ArrayList<>();
      ducks.add(new MallardDuck());
      ducks.add(new RocketDuck());
      ducks.add(new RubberDuck());

      for (Duck duck : ducks) {
         duck.display();
         duck.fly();
         duck.quack();
         System.out.println("\n");
      }

      System.out.println("Poor rubber ducky can't fly. Let's change this behaviour by strapping on a rocket!");

      Duck rubber_ducky = ducks.get(2);
      rubber_ducky.setFlyBehavior(new FlyRocketPowered());
      rubber_ducky.display();
      rubber_ducky.fly();

   }
}
